### DISCUSSION TITLE
Very Awesome Idea for Language syntax simplification

### DISCUSSION THREAD

**Discussion by SlimeNull** [dotnet/csharplang#5015] [Language Ideas]
# C# syntax suggestions

## English
1. use 'group block' for members:
    ```csharp
   class MyClass {
       // we are using syntax like this for public members:
       public string Age;                 // Test Field
       public string Name { get; set; }   // Test Property
       public void PrintName() {          // Test Method
           Console.WriteLine(Name);
       }
       
       // default is private is no modifier specifed
       string testPrivateField;
       string printPrivateField() {
           Console.WriteLine(testPrivateField);
       }
       
       // and i hope there is a syntax like this:
       public {                       // all members in 'public' block is public
           string Age;
           string Name { get; set; }
           void PrintName(){
               // statements
           }
       }
       private {
           // and 'private' is the same
       }
       protected {
           // 'protected' also
       }
       
       // this is just Syntactic sugar
       
       // and,,, maybe this?
       [AnyAttribute]
       {
           // some member, field, property, or method
       }
       
       // or this?
       static {
           // some static members
       }
       
       // and this?
       public static {
           // some public and static members
       }
   }
    ```

2. use auto Type judgement for variable, property, and method

   ```csharp
   class MyClass {
       // now we can use var for auto Type judgement:
       var myInt = 12312;
       // but we cannot use multiple variable declaration:
       var
           myInt1 = 234,
           myInt2 = 34534;    // i hope this can be a correct syntax
       
       // and this syntax:
       var myStr;   // to get the Type of this field,
                    // compiler should find the reference,
                    // and find the first assignment for this variable
                    // if there is no assignment and other using, then not compile
                    // this variable, and give a warning. if somewhere try get the
                    // value before assignment for this variable, give a error,
                    // and the variable should not be assigned with multiple types,
                    // such as that assign it with 123 at method A, and assign it with
                    // "this is a string" at method B.
       // also, this syntax should be supported:
       var myStr1, myFloat, myDouble;
       
       // and also, property:
       var myProp { get; set; }
       
       // assignment for variables and properties above.
       void TestAssignment() {
           myStr = "string value";
           myStr1 = myStr;
           myFloat = 123.3f;
           myDouble = 345345.65d;
           myProp = myFloat + myDouble;    // result is a double float number
       }
       
       // and auto return value Type judgement:
       func myFunc(int a, int b) {   // use func for auto judgement
           return a + b;   // return an integer
       }
   }
   ```

3. simplify syntax for same Type parameters:

   ```csharp
   class MyClass {
       double Sum(double a, double b)    // this is a method
       {
           return a + b;
       }
       
       // simplified syntax:
       double Sum(double a, b)     // because b is also 'double' Type, 
                                   // so, ignore the Type modifier
       {
           return a + b;
       }
       
       // and also:
       string SumJoin(int a, b, string str1, str2)
       {
           int count = a + b;
           string str = str1 + str2;
           StringBuilder sb = new();
           for (int i = 0; i < count; i++)
               sb.Append(str);
           return sb.ToString();
       }
   }
   ```

## Chinese 中文

1. 对成员使用 '群块':
    ```csharp
   class MyClass {
       // 对于公共的成员我们现在正在使用的是这种语法:
       public string Age;                 // 测试字段
       public string Name { get; set; }   // 测试属性
       public void PrintName()
       {          // 测试方法
           Console.WriteLine(Name);
       }
       
       // 如果没有修饰符指定, 那么默认是私有的
       string testPrivateField;
       string printPrivateField()
       {
           Console.WriteLine(testPrivateField);
       }
       
       // 我希望能有这种语法:
       public
       {                       // 所有在 'public' 块中的成员都是公共的
           string Age;
           string Name { get; set; }
           void PrintName(){
               // 语句
           }
       }
       private
       {
           // 'private' 也是一样
       }
       protected
       {
           // 'protected' 也是
       }
       
       // 当然, 这只是些语法糖
       
       // 然后, 还有这样?
       [AnyAttribute]
       {
           // 某些成员, 字段, 属性, 或者方法
       }
       
       // 或者这样?
       static
       {
           // 某些静态成员
       }
       
       // 然后这样?
       public static
       {
           // 某些公共静态成员
       }
   }
   ```

2. 对字段, 属性以及方法使用自动的类型判断
   ```csharp
   class MyClass {
       // 现在我们可以用 var 来实现自动的类型判断
       var myInt = 12312;
       // 但是这种多个变量的赋值就不彳亍了
       var
           myInt1 = 234,
           myInt2 = 34534;    // 我希望这种语法能受支持
       
       // 然后是这种语法:
       var myStr;   // 获取这个字段的类型,
                    // 编译器应该去寻找这个字段的引用,
                    // 然后查找它的第一个赋值语句,
                    // 如果没有赋值语句或者其他使用, 那就不编译这个变量
                    // 然后给一个警告, 如果某些地方试图在变量赋值前就
                    // 获取这个变量的值, 则给出一个错误
                    // 并且这个变量也不应该有多个类型的赋值
                    // 例如在 A 方法中为他赋值 123, 在 B 方法中为他
                    // 赋值 "这是个字符串"
       // 并且这种语法也应该受支持
       var myStr1, myFloat, myDouble;
       
       // 还有属性
       var myProp { get; set; }
       
       // 上面的变量, 属性的赋值语句
       void TestAssignment() {
           myStr = "string value";
           myStr1 = myStr;
           myFloat = 123.3f;
           myDouble = 345345.65d;
           myProp = myFloat + myDouble;    // 结果是一个双精度浮点数
       }
       
       // 以及自动的返回值判断:
       func myFunc(int a, int b) {   // 使用 func 来自动判断
           return a + b;   // 返回一个整数
       }
   }
   ```

3. 简化同样类型的参数语法

   ```csharp
   class MyClass {
       double Sum(double a, double b)    // 这是个方法
       {
           return a + b;
       }
       
       // 简化的语法:
       double Sum(double a, b)     // 因为 b 也是 double 类型, 
                                   // 所以就忽略这个类型修饰符吧
       {
           return a + b;
       }
       
       // 还有这样:
       string SumJoin(int a, b, string str1, str2)
       {
           int count = a + b;
           string str = str1 + str2;
           StringBuilder sb = new();
           for (int i = 0; i < count; i++)
               sb.Append(str);
           return sb.ToString();
       }
   }
   ```

**Comment by sharwell** [comment]
1. This is covered by https://github.com/dotnet/csharplang/discussions/1021
2. This is covered by https://github.com/dotnet/csharplang/issues/601
3. I couldn't find another issue for this request. It may be best to file it as a separate request if no one else identifies a duplicate.
4. I couldn't find another issue for this request. It may be best to file a separate request, but that request will need to address the manner in which breaking changes are avoided.

**Reply by SlimeNull** [reply]
for c# style syntax，we should use "{}", but not c艹 style like ":", because ':' means 'extends' or 'implement' in c#, that's I think.

**Comment by HaloFour** [comment]
4. Sounds like it wants to imitate the behavior of nested classes in Java, where non-static nested classes implicitly inherit an instance of the parent class.  Trying to adopt Java syntax for this would be a breaking change as normal nested classes in C# already behave as static nested classes in Java and converting them to be non-static would change their contract and behavior in a breaking way.  A new syntax would be required to denote that the nested class does this.

IMO, it's not necessary to have a language feature for this behavior.  If you want the behavior of a non-static class you can pass an instance of the parent class to the constructor manually.  The nested class has privileged access to the private members of the parent class.

Further, the suggested syntax doesn't even work in Java as there is no implicit reference of `MyClass` to pass to the synthesized constructor of `MyClass.InClass1`.  What instance of `MyClass` would you be referencing from `MyClass.InClass1`?

**Reply by SlimeNull** [reply]
See the code, I used this:
Instance.Classname varName = new();
Of course the instance~

But alright, it's not necessary, but I think if c# can use java style 'non-static' class, the language will be better, and for variables and other members in class, static is also means 'owned by the class self' and 'non-static' means 'owned by the instance', I just want to make the syntax be more unitive, by the way, in vb.net, all classes can be instantiated。

and i now, it's a very big breaking change for c#(old projects), but I still think it's a good change.

If these changes above are applied, c# will be one of the most perfect programming language designs

**Reply by theunrepentantgeek** [reply]
> it's a very big breaking change for c#(old projects), but I still think it's a good change.

This is itself a very good reason why it should never happen - go look at the pain the Python community went through. Python 3 was introduced in 2008 but it took until 2020 - twelve years! - before they were able to retire Python 2, and even then there were enormous complaints.

You haven't described at all the problem you're trying to solve. 
What's wrong, and why is this solution the best way to solve those problems.

Why is it worth breaking billions of lines of existing C# code (yes, billions with a B) and forcing all those development teams to make extensive changes to the code they already have - code that's working - in order to accommodate these changes?

**Reply by SlimeNull** [reply]
Just suggestion~~~
Maybe we can add a compiler option for this feature?

**Reply by Joe4evr** [reply]
That would introduce a dialect to the language, [and we just had a thread where it's explained that that's not something the C# language team wants to pursue without severe justification](https://github.com/dotnet/csharplang/discussions/5012).

**Reply by HaloFour** [reply]
I see.  The difference in case makes that difficult to see.  You can accomplish the same behavior today by doing the following:

[SharpLab](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQHMIDsIIJYDGqsM+wcMKANDCAnFgD4ACATAIwCwAUD8wMwACNoIDyFXAGEANlGTJBAbx6DVggA74AbrAiCkUACYB7LNICew9gAZBWKAFsIAbhVq3qgWIkIZc5AAUzDZ2jhAAlIIAvAB8gjAAFnjIAHT2TtGhTq7caoIewkIiAJJYOL6y8koFeZp4OpT60CZmluKUFf4aUEhYMDl5eTVqXqXlfvIB7VKVCuo92DCRsfFJqfO9MJkbiwODqsOeRQAsggCyAZHKufuDwQCcAQAkAEQAEhDS0sZKO31pYQAvgBCF7hPb7QEFKHcGF8dgANmErCsAHZqjdPIjhKczlA8FhLhjbt4OhMFMYfJkcAB3AIvADqxgQ0kMYIhg2mnXkKTGMy6BPK1IgdMpHXBh0EgtwKQuEsxghhgKAA==)

It's a little more verbose but I don't think nested classes are so commonly used to justify it.  Even in Java you can't instantiate a nested class in that manner, you can only do so from an instance member of the outer class.

**Reply by HaloFour** [reply]
> If these changes above are applied, c# will be one of the most perfect programming language designs

I get the sentiment but ask 100 devs what features would make C# the most perfect programming language and you'll get 1000 different answers.  :)

**Comment by ghost** [comment]
针不戳

但是我总是觉得"简化同样类型的参数语法"怪怪的
尤其是
```csharp
string SumJoin(int a, b, string str1, str2)
```
看起来很难受(我是个强迫症)

**Reply by SlimeNull** [reply]
草草草，什么风把您给刮来了！

**Reply by SlimeNull** [reply]
但是不可否认的是，这的确能减少你手部负担

**Reply by ghost** [reply]
哈哈哈

**Reply by ghost** [reply]
确实

**Reply by ghost** [reply]
就是说有没有什么更好的方案代替一下 不然我是真的抓狂

**Reply by SlimeNull** [reply]
你可以添加换行，然后看起来就好多了
```csharp
string method(
    int a, b,
    string str1, str2)
{
    // Code here
}
```

**Reply by ghost** [reply]
wow~ ⊙o⊙ 可以可以

**Comment by huoyaoyuan** [comment]
Some short answers based on my personal knowledge of the compiler and the team:

1. It is benefitial, but not significantly. It is somehow unfriendly for IDE code fixes to midify the code.

2. This is **explicitly denied** by the design of C#. The signature of public members must be explicitly specified.
This will also degrade the performance of compiler.

3. The benifit is also too small, and also unfriendly for tooling.

4. This should be a new kind of type nesting. Quite useful, but we can't modify existing syntax.

**Reply by SlimeNull** [reply]
these suggestions are based golang c艹，and Java。such as 'group block'，you know，c艹 don't have to specify 'public' for every public member. the forth suggestion is not necessary, and unreasonable. But the first three suggestions are very good, and I think it's not hard to do that for compiler, because compiler do check member references.

For the third suggestion, I think it can exist as a extension syntax, user can also use current syntax。

They are just syntactic suger.

Maybe you think code's readability will decrease, but look at c
艹，if you mean the simplified syntax of same-type parameters, just look at golang.

User may not use these, but it's important, I think.

If your project is very big, don't is troublesome to specify 'public' for every member?

**Reply by huoyaoyuan** [reply]
Yes, the first two suggestions are useful. However, *every simple language change is quite costful, including changes in IDE, analyzers and other tooling* (quoted from official member). It has to be significantly beneficial for a "simplification" change.

**Reply by huoyaoyuan** [reply]
> For the third suggestion, I think it can exist as a extension syntax, user can also use current syntax

It has been explicitly denied for many times.

> User may not use these, but it's important, I think.

The C# team do not tend to give user too many ways to do the same thing. They also avoid over simplification because it may block future changing room of the language.

Syntax sugars are easily to be harmful to language parser. The complexity of parser is also considered in language design.

**Reply by spydacarnage** [reply]
The problem with the 1st suggestion, in my opinion, is that if I see this line in a file:
```cs
void PrintName() { ... }
```
I then have to potentially scroll up the file just to check whether it is inside a `public` or `protected` block, or whether it is just using the default `private`.  Syntax sugar which makes code harder to read and understand is a massive negative in my book.

**Reply by theunrepentantgeek** [reply]
> If your project is very big, don't is troublesome to specify 'public' for every member?

No. Not at all. Emphatically not. 

As someone who has been working extensively in **Go** over the last couple of years, what's troublesome is having to remember an arcane rule about letter case that controls visibility - a role that's different from **_every_** other language I've used over the last forty years of programming. 

What's troublesome is only having the equivalent of **public** and **internal**, making it difficult to achieve rigorous encapsulation. 

What's troublesome is that a refactoring to change method visibility permeates throughout the code. 

The Go language has some nice features, but this is not one of them.

**Comment by RikkiGibson** [comment]
I just wanted to say I appreciated the title of the post. 😄

However, I do not know if any of the suggestions will be picked up. Like @sharwell said you might want to consider opening a standalone discussion for (3).

(4) is indeed just the way Java does inner classes, and they reserved 'static class' to mean that an inner class doesn't hold an instance of the containing class. I think it's not likely the language team would want to introduce special support for this for reasons @HaloFour outlined.

**Reply by SlimeNull** [reply]
Thanks for your reply, I'll do that for (3), and I really think that's very useful, we could type less code with this syntax~~~
